### DISCUSSION TITLE
Replacement for OpenTK 3.x' `GLControl` (render large framebuffers on a canvas performantly)?

### DISCUSSION THREAD

**Discussion by YoshiRulz** [unoplatform/uno#8851] [Q&A] [answered]
The main functionality of my app is rendering a framebuffer (which comes from a P/Invoke'd lib).
My initial thought when looking through Uno's control inventory was swapping out .bmps on an `Image`. Helpful people on Discord told me I could use the underlying Skia instead.
Unfortunately, the [sample app](https://github.com/unoplatform/Uno.Samples/tree/5877432bdaead673921113a0798a273fb1a0f6e5/UI/SkiaSharpTest) doesn't work, it crashes with `The version of the native libSkiaSharp library (80.3) is incompatible with this version of SkiaSharp` (mono/SkiaSharp#1715). Taking a guess that the version printed was synced to SkiaSharp versions, I downgraded the latter to 2.80.4, which worked.
But the app now whitescreens with `MissingMethodException: Method not found: 'UIAsyncOperation CoreDispatcher.RunAsync(CoreDispatcherPriority, DispatchedHandler)` (mono/SkiaSharp#1877).

Where should I go from here? Do I downgrade Uno to 3.x? For a PoC, the only requirements are that it targets `netstandard20`/`net48` x64 for Windows and Linux (via Mono).

Background:
I'm modernising the FOSS emulator project [BizHawk](https://github.com/TASEmulators/BizHawk). The emulator frontend is in WinForms, targeting `net48`. The backend is a collection of `netstandard20` libraries, native unmanaged libraries (most Win+Linux, some Win), and ELF binaries (long story).
The critical emulation functionality is that the frontend passes execution to one of the libraries until it reaches a VBlank, then it returns a width, height, and pointer/reference to an int array (just an RGBA value for each pixel). Currently, the framebuffer is drawn on a WinForms `UserControl` with Direct3D9 (SlimDX) when running on Windows, or else drawn with OpenTK on its `GLControl`. (There's also a GDI+ renderer but it's too slow to use—I'm hoping Skia doesn't have that problem.)
I've been slowly moving code to `netstandard20` libraries and swapping out dependencies in preparation for updating the Windows frontend to .NET 6 WinForms, but that would mean leaving Linux behind with Mono WinForms. Working with WinForms is painful enough on its own; I'm not keen on juggling two versions. So, with no MAUI on Linux either, of the third-party options I chose Uno.

**Comment by jeromelaban** [comment, marked as answer]
Thanks for posting your question here. ~~So far, there's no stable version of SkiaSharp which supports Uno 4.x, and you'll have to bump to SkiaSharp experimental builds. Can you try and let us know? Let's discuss the rest once you get it running.~~  (Latest stable builds of SkiaSharp support Uno 4.x)

**Reply by jeromelaban** [reply]
Also, make sure to try with Uno 4.3.x stable builds, we made some corrections there as well.

**Reply by YoshiRulz** [reply]
I already updated Uno packages to 4.3.5. With `SkiaSharp.Views.Uno` 2.88.0 from yesterday I'm still running into mono/SkiaSharp#1715.

**Reply by jeromelaban** [reply]
Thanks. Could you try creating a minimal repro sample which fails, and attach it here?

**Reply by YoshiRulz** [reply]
It's the Skia canvas sample https://github.com/YoshiRulz/Uno.Samples/tree/93f9a1cbd0b4688868ca7f16210d80ef227b3c1b/UI/SkiaSharpTest

**Reply by jeromelaban** [reply]
I can run the sample you've extracted just fine (@93f9a1cbd0b4688868ca7f16210d80ef227b3c1b). Note that under Uno Skia+Gtk, the "hardware accelerated" sample does not apply, since the underlying SKXamlCanvas rendering uses what the GTK Uno backend uses. It is configured by `GtkHost.RenderSurfaceType` and defaults to OpenGL or OpenGL ES depending on the platform.

Can you provide the exact app log output you're getting on your end?

**Reply by YoshiRulz** [reply]
stdout: https://gist.github.com/YoshiRulz/c373e456ee985d844cb80728a81aff79
Are you on Windows? I am using Rider on Linux, though the above was taken with the dotnet CLI.

**Reply by jeromelaban** [reply]
Thanks indeed, this makes more sense.

When working with SkiaSharp, you'll need to align the "NativeAssets" packages with the SkiaSharp version you're using. In this case, you're using "2.88.0", which is higher than the version referenced by Uno itself (See the dependencies tab of [Skia.Gtk support package](https://nuget.info/packages/Uno.UI.Runtime.Skia.Gtk/4.3.5)). 

In short, to fix this, add the following:

```xml
<PackageReference Include="SkiaSharp" Version="2.88.0" />
<PackageReference Include="SkiaSharp.NativeAssets.Linux" Version="2.88.0" />
```

We'll be updating our docs to explain this better.

**Reply by YoshiRulz** [reply]
Ah! That did it, thanks. I'll start on hacking emulation into this tomorrow.

**Reply by YoshiRulz** [reply]
Immediate roadblock: `Uno.UI.Runtime.Skia.Gtk` isn't packaged for `net461`, though its dependencies are all targeted lower. I quickly tried building it in GitPod but got lost.

**Reply by jeromelaban** [reply]
We're not considering adding support for net461 for that target. Why is net6 not possible for you?

**Reply by YoshiRulz** [reply]
I suppose it is possible in my case, just more annoying.

**Reply by jeromelaban** [reply]
Why is it annoying? net6 support on Linux is a lot more stable than what mono/net46 was in most cases. If you were having Windows 7 as a target, you could use the WPF head, but maybe it's not a concern for you.

**Reply by YoshiRulz** [reply]
We still have a few non-optional dependencies which are only available for .NET Framework, chiefly OpenTK 3.x.

**Reply by jeromelaban** [reply]
I see. What's the use for OpenTK? The current structure for an Uno App does not make it easy to use it. Can you provide an example of what you're trying to do with it?

**Reply by YoshiRulz** [reply]
As I wrote above, it's only rendering a rect on a WinForms control, not anything 3D to my knowledge. The code is buried under abstraction spaghetti but most of the OpenGL calls are in [this file](https://github.com/TASEmulators/BizHawk/blob/1026503d9241203987dc695a01e4907dcaaffb1f/src/BizHawk.Bizware.OpenTK3/IGL_TK.cs). Apart from graphics, we use it for audio and input on Linux. I'm considering [SharpAudio](https://github.com/feliwir/SharpAudio) and [evdev-sharp](https://github.com/afshin-parsa/evdev-sharp) as replacements.

**Reply by jeromelaban** [reply]
Got it. We could try adding the net461 or net48 TargetFramework to see if this helps. Everything's in this package: https://github.com/unoplatform/uno/blob/233001f91d618e4841a971d179da0fa873416879/src/Uno.UI.Runtime.Skia.Gtk/Uno.UI.Runtime.Skia.Gtk.csproj#L4.
